File: 2021/05/31/main.go
package main

import (
	"sort"
	"strings"
)

func suggestedProducts(products []string, searchWord string) [][]string {
	sort.Strings(products)
	var a [][]string
	for i := 1; i <= len(searchWord); i++ {
		q := searchWord[:i]
		j := 0
		var st []string
		for j < len(products) {
			p := products[j]
			if strings.HasPrefix(p, q) {
				st = append(st, p)
			}
			j += 1
			if len(st) == 3 {
				break
			}
		}
		a = append(a, st)
	}
	return a
}

File: 2021/05/31/main.py
class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        a = []
        def bs(q, d):
            l, h = 0, len(products)
            while l < h:
                m = l + (h - l) // 2
                p = products[m]
                if q <= p[:d]:
                    h = m - 1
                else:
                    l = m + 1
            return l
        for i in range(len(searchWord)):
            q = searchWord[:i+1]
            d = len(q)
            pos = bs(q, d)
            st = []
            while pos < len(products):
                p = products[pos]
                if q == p[:d]:
                    st.append(p)
                pos += 1
                if len(st) == 3:
                    break
            a.append(st)
        return a